activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -453,6 +453,9 @@ def checkout_and_verify(c)
           c.verify!
         end
         c
+      rescue
+        disconnect!
```

Comment:
Yes, disconnect! will disconnect all active connections.
And no, exception does not raise only when new connection fails, but raises when active connection is disconnected. And, it means shutdown of database(or network connectivity), so anyway all the connection is closed.